### PR TITLE
[angular-typescript] fix using form data

### LINKED ISSUE #6457: [Typescript-Angular2] Setting multipart/form-data will always send [object FormData] as body
##### Description

At the moment it is not possible to actually send multipart/form-data requests using generated code with the typescript-angular2 generator as it will always call formParams.toString() for the body contents which will always result in [object FormData] in the actual request.

##### Swagger-codegen version

v2.3.0

##### Swagger declaration file content or url

Problematic file: [swagger-codegen/modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache](https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache#L185)

Problematic lines:
```mustache
let formParams = new (useForm ? FormData : URLSearchParams as any)() as {
   set(param: string, value: any): void;
};
...
{{#hasFormParams}}
    body: formParams.toString(),
{{/hasFormParams}}
```

##### Steps to reproduce

1. Create an API service which can upload files
2. Generate the API client as usual using typescript-angular2 template
3. Try to use the Service with a file
=> Request will show up with [object FormData] in the body

##### Suggest a fix/enhancement

Angulars HTTP handling of forms is actually pretty smart. I think simply changing the line to the following works.

```mustache
{{#hasFormParams}}
    body: useForm ? formParams : formParams.toString(),
{{/hasFormParams}}
```

### PR DESCRIPTION
**Pull request by macjohnny**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [x] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming langauge.

### Description of the PR
fixes several issues:
* fixes #6457 `Setting multipart/form-data will always send [object FormData] as body`
* re-apply fix for #6306 that went missing with #6295 and adapt it for usage with `HttpClient`
* `FormData#set is unable to use in Edge/IE, so we should use FormData#append instead.`
* add missing `bin/windows/typescript-angular-v4.3-with-npm.bat`
* make ` bin/typescript-angular-v4.3-with-npm.sh` executable
* remove unused method `isJsonMime()`

As far as I have tested it on Edge, Chrome, it seems to work fine. I also tested the `v4.0` and `v4.3` clients. The `v2.0` client is tested on CI.

### CODE REVIEW

**Comment by macjohnny** [comment]
cc @roni-frantchi @wing328 @sebastianhaas

File: samples/client/petstore/typescript-angular-v4/npm/package.json [hunk 1]
```diff
@@ -35,6 +35,6 @@
     "typescript": "^2.1.5"
   },
   "publishConfig": {
-    "registry": "https://skimdb.npmjs.com/registry"
```
Comment (sebastianhaas) [inline samples/client/petstore/typescript-angular-v4/npm/package.json]: Do you know where these non-related changes in the generated output comes from?

File: samples/client/petstore/typescript-angular-v4/npm/api/pet.service.ts [hunk 2]
```diff
@@ -274,7 +274,7 @@ export class PetService {
      */
     public deletePetWithHttpInfo(petId: number, apiKey?: string, extraHttpRequestParams?: any): Observable<Response> {
         const path = this.basePath + '/pet/${petId}'
-                    .replace('${' + 'petId' + '}', String(petId));
+                    .replace('${' + 'petId' + '}', encodeURIComponent(String(petId)));
```
Comment (sebastianhaas) [inline samples/client/petstore/typescript-angular-v4/npm/api/pet.service.ts]: Do you know where these non-related changes in the generated output comes from? Have you pulled the latest changes before running your enhancements? I think this `encodeURI` thing was introduced in another PR and therefore should be already present?

File: modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache [hunk 3]
```diff
@@ -274,7 +274,8 @@ export class {{classname}} {
             body: {{paramName}} == null ? '' : JSON.stringify({{paramName}}), // https://github.com/angular/angular/issues/10612
 {{/bodyParam}}
 {{#hasFormParams}}
-            body: formParams.toString(),
+            // https://github.com/swagger-api/swagger-codegen/issues/6457
```
Comment (sebastianhaas) [inline modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache]: I'm not sure about that reference. Sometimes there are references to unsolved issues on Angular-side that need to be addressed at some point, but in this case it's just a proper implementation we haven't had before so I don't think that reference is really useful or required. If someone looks up the changes, they will find this PR anyways.

What do you think @TiFu @taxpon @kenisteward ?

File: samples/client/petstore/typescript-angular-v4/npm/package.json (hunk 1 above)
Comment (macjohnny) [reply samples/client/petstore/typescript-angular-v4/npm/package.json]: that's a good question, I also noticed this unrelated change, but I guess this comes from a merged change to the master without running the bin/windows/typescript-angular-v4-with-npm.bat script.

File: samples/client/petstore/typescript-angular-v4/npm/api/pet.service.ts (hunk 2 above)
Comment (macjohnny) [reply samples/client/petstore/typescript-angular-v4/npm/api/pet.service.ts]: this comes from the fact that in https://github.com/swagger-api/swagger-codegen/pull/6525 I did run the `bin/windows/typescript-angular2-petstore-all.bat` script, but forgot to run the `bin/windows/typescript-angular-v4-with-npm.bat` script. This shows that the CI tests somehow did not check the angular v4 output samples. Maybe this needs to be fixed? @sebastianhaas will you do that?

File: modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache (hunk 3 above)
Comment (macjohnny) [reply modules/swagger-codegen/src/main/resources/typescript-angular/api.service.mustache]: @sebastianhaas yes, you are right, with git blame one can easily find the last change of that line, so I will remove the comment

**Comment by macjohnny** [comment]
@sebastianhaas please check again, the reference to the issue was removed from the code
@wing328 this should be merged before #6295

**Comment by macjohnny** [comment]
while verifying this PR, I descovered another issue: #6588

**Comment by macjohnny** [comment]
@sebastianhaas how can I re-run the CI check?

**Comment by wing328** [comment]
I just restarted it. Let's see how it goes.

(another way is to close and then reopen the PR)

**Comment by macjohnny** [comment]
@wing328 the check passed. could you have a look at it and merge it?

**Comment by wing328** [comment]
cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx for review

**Comment by macjohnny** [comment]
@wing328 this should not be merged before #6295, I will resolve the conflicts as soon as #6295 is merged into master.

**Comment by wing328** [comment]
@macjohnny #6295 has been merged into master and I've applied some minor fixes. Please pull the latest master to rebase your PR. Thanks!

**Comment by macjohnny** [comment]
@sebastianhaas @wing328 @bedag-moo @www2k please review.

See also https://github.com/swagger-api/swagger-codegen/pull/4623

There is one `TODO` at https://github.com/swagger-api/swagger-codegen/pull/6574/files#diff-5bdae26ba753c4986ceb5af35bada047R224 due to an already existing bug. any idea how to elegantly solve it?

**Comment by macjohnny** [comment]
@wing328 could you please restart the CI jobs?

**Comment by wing328** [comment]
Done. Another way is to close the PR and then reopen it.

**Comment by macjohnny** [comment]
@sebastianhaas are there any plans for an integratin test verifying the functionality of the generated angular code in an angular project?

**Comment by macjohnny** [comment]
cc @krizic @taxpon @kenisteward

**Comment by watiko** [comment]
This comment is not directly related to this PR.

AFAIK `FormData#set` is unable to use in Edge/IE, so we should use `FormData#append` instead.

**Comment by wing328** [comment]
> AFAIK FormData#set is unable to use in Edge/IE, so we should use FormData#append instead.

@watiko Can you please open an issue to track it?

**Comment by macjohnny** [comment]
@watiko thanks for your comment, you are absolutely right, IE / Edge does not support `FormData.set()`. I updated the PR accordingly. Could you please also test / review it?

**Comment by macjohnny** [comment]
@taxpon can you review and verify I correctly re-applied your fix for https://github.com/swagger-api/swagger-codegen/issues/6306 that went missing with #6295 ?

**Comment by macjohnny** [comment]
@roni-frantchi @wing328 @sebastianhaas @taxpon @watiko @krizic @taxpon @kenisteward @TiFu  @Vrolijkx anyone up for reviewing and testing this PR? 

it fixes several issues:
* #6457 `Setting multipart/form-data will always send [object FormData] as body`
* re-apply fix for #6306 that went missing with #6295 and adapt it for usage with `HttpClient`
* `FormData#set is unable to use in Edge/IE, so we should use FormData#append instead.`
* add missing `bin/windows/typescript-angular-v4.3-with-npm.bat`
* make ` bin/typescript-angular-v4.3-with-npm.sh` executable
* remove unused method `isJsonMime()`

As far as I have tested it on Edge, Chrome, it seems to work fine. I also tested the `v4.0` and `v4.3` clients. The `v2.0` client is tested on CI.

**Comment by macjohnny** [comment]
@bedag-moo @sebastianhaas @taxpon @krizic @wing328 I kindly ask you to review this pull request, as it fixes several issues with the typescript/angular2/4 swagger codegen template.

**Comment by macjohnny** [comment]
@sebastianhaas thanks a lot for reviewing!
@wing328 how about merging this PR?

**Pull request merged**